R/guess_provider.R

```diff
@@ -46,26 +24,39 @@ guess_provider <- function(pkg, verbose = FALSE) {
   }
 }
 
-#
-# Helper functions can later be moved to utils.R
-#
-
 # available_source_packages ----------------------------------------------------
-available_source_packages <- function(url) {
+codemeta_cache_env <- new.env(parent = emptyenv())
+
+available_source_packages <- function(repo = c("CRAN", "BIOC")) {
+
+  url <-
+    switch(
+      repo,
+      CRAN = "https://cloud.r-project.org",
+      BIOC = "https://www.bioconductor.org/packages/release/bioc",
+      stop("Only CRAN and BIOC repos are supported.")
+    )
+
+  contrib_url <- utils::contrib.url(url, "source")
+
+  if (is.null(codemeta_cache_env[[repo]])) {
+    codemeta_cache_env[[repo]] <- utils::available.packages(contrib_url)
```

Comment:
wow, doing a direct assign into an `env` is new to me; I'm used to using `get` and `assign` explicitly when working with environments.  I dunno if there's any established best practice to prefer one method over the other?  I guess `get` and `assign` are mostly for when you want to look up the env by name, but here we can just ref the env object directly?

Reply:
>  I dunno if there's any established best practice to prefer one method over the other?

I am not entirely sure re. established best practice, but I used [this r-hub post](saving-results-in-an-environment) as a starting point if that is any indication. With the slight adjustment of using `codemeta_cache_env[[repo]]` instead of codemeta_cache_env$repo because of (non-)standard evaluation. '

Funny enough, though I am vaguely aware of `get()` and `assign()`, I never use it.

Maybe @maelle knows more?

Reply:
not sure whether one is best, to be honest. :thinking: